server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -0,0 +1,230 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cloudera.livy.utils
+
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeoutException
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent._
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+import org.apache.hadoop.yarn.api.records.{ApplicationId, ApplicationReport, FinalApplicationStatus, YarnApplicationState}
+import org.apache.hadoop.yarn.client.api.YarnClient
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+
+import com.cloudera.livy.{LivyConf, Logging, Utils}
+import com.cloudera.livy.rsc.{Utils => RSCUtils}
+import com.cloudera.livy.util.LineBufferedProcess
+
+object SparkYarnApp extends Logging {
+  // Create a non daemon thread pool to run getAppIdFromTag().
+  // getAppIdFromTag() might take a while and should not use the global default thread pool.
+  private implicit val ec = ExecutionContext.fromExecutor(
+    Executors.newCachedThreadPool(RSCUtils.newDaemonThreadFactory("getAppIdFromTag-%d")))
+
+  // YarnClient is thread safe. Create once, share it across threads.
+  lazy val yarnClient = {
+    val c = YarnClient.createYarnClient()
+    c.init(new YarnConfiguration())
+    c.start()
+    c
+  }
+
+  def fromAppTag(
+      appTag: String,
+      process: Option[LineBufferedProcess],
+      listener: Option[SparkAppListener],
+      livyConf: LivyConf): SparkYarnApp = {
+    val deadline = getYarnTagToAppIdTimeout(livyConf).fromNow
+    val pollInterval = getYarnPollInterval(livyConf)
+    new SparkYarnApp(getAppIdFromTag(appTag, pollInterval, deadline), process, listener, livyConf)
+  }
+
+  /**
+   * Find the corresponding YARN application id from an application tag.
+   *
+   * @param appTag The application tag tagged on the target application.
+   *               If the tag is not unique, it returns the first application it found.
+   *               It will be converted to lower case to match YARN's behaviour.
+   * @return ApplicationId or the failure.
+   */
+  @tailrec
+  private def getAppIdFromTag(
+      appTag: String,
+      pollInterval: Duration,
+      deadline: Deadline): ApplicationId = {
+    val appTagLowerCase = appTag.toLowerCase()
+
+    // FIXME Should not loop thru all YARN applications but YarnClient doesn't offer an API.
+    // Consider calling rmClient in YarnClient directly.
+    val appType = Set("SPARK").asJava
+    yarnClient.getApplications(appType).asScala.find(_.getApplicationTags.contains(appTagLowerCase))
+    match {
+      case Some(app) => app.getApplicationId
+      case None =>
+        if (deadline.isOverdue) {
+          throw new Exception(s"No YARN application is tagged with $appTagLowerCase.")
+        } else {
+          Clock.sleep(pollInterval.toMillis)
```

Comment:
My bad. In that case you don't need the executor context instantiated in this object, do you?

In general I think this code is a little too complicated and hard to follow. I've thought about some ways to make it easier, my current preferred way is:

- have a base class with an abstract `def getAppId(): String`
- have an implementation that has a known app id defined in the constructor (I think you don't need that right now?)
- have a separate implementation where that method's implementation does the equivalent of:

```
override def getAppId(): String = {
  val exitCode = process.waitFor()
  if (exitCode != 0) {
    throw some exception 
  }

  // add code that maps the tag to the app id here
}
```

The base class's `yarnAppMonitorThread` would call `getAppId()` as the first thing it does. You can use a promise for that so that you can keep the current `kill()` implementation. This way the interface becomes cleaner - the non-tag version of the class doesn't need to know about processes and what not, and it's easier to figure out the sequence of actions.

I'm not 100% happy with that, though, so open to suggestions.